### ISSUE TITLE
[🐞] <Link> Fallback to Hard Navigation when Page's JS file is not found.

### ISSUE THREAD

**Issue by iamriajul** [BuilderIO/qwik#5514]
### Which component is affected?

Qwik City (routing)

### Describe the bug

Expect a scenario where You've deployed the application initially and your user has visited your home page.
The user was sitting down a while after opening the homepage, In the meanwhile you changed some code to your Login page, and now what happens is that your js bundle's filename for the login page components has changed to something else (because of hash-calculated based on code). 
At this moment when the user tries to navigate to the Login page via the Link component (using Soft Navigation), the navigation just hangs! The user doesn't understand what is happening and why the website is taking so long, but we know if the user waits infinitely the page won't load because the js filename has changed, to reflect the change user must hard reload.
This is the scenario we want to fix by falling back to the old good reliable MPA / Hard Navigation.

### Reproduction

https://github.com/iamriajul/repro-fallback-to-hard-navigation-when-page-js-not-available

### Steps to reproduce

- `pnpm install`
- `pnpm build`
- Now delete the file: `/build/q-ef638c7f.js` (to simulate you have pushed a new build with code changes, which in turn changes the filename.)
- `pnpm serve`
- Visit: `http://localhost:3000`
- Scroll Down and Click [Todo App](http://localhost:3000/demo/todolist) Link. Now see what happens! nothing happens the application just silently crashes.

### System Info

```shell
System:
    OS: macOS 13.5.1
    CPU: (16) x64 8-Core Processor             
    Memory: 988.09 MB / 32.00 GB
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 16.20.2 - ~/.nvm/versions/node/v16.20.2/bin/node
    Yarn: 1.22.19 - ~/.nvm/versions/node/v16.20.2/bin/yarn
    npm: 8.19.4 - ~/.nvm/versions/node/v16.20.2/bin/npm
    pnpm: 8.9.2 - ~/.nvm/versions/node/v16.20.2/bin/pnpm
  Browsers:
    Chrome: 119.0.6045.105
    Edge: 119.0.2151.97
    Safari: 16.6
```

**Comment by gioboa** [comment]
Thanks for your feedback. Do you have a possibile solution on top of your head?

**Comment by iamriajul** [comment]
I've been able to fix the issue like this:
![image](https://github.com/BuilderIO/qwik/assets/24244379/ff096123-7d52-472e-840b-54ca95622325)

**Comment by knpwrs** [comment]
I was just wondering about this myself. What happens when a new version is deployed and the build hash changes, but the user is on a stale page and tries to interact with stuff?

Side note: this isn't a contrived issue. I've seen this both in my full time job (vite/react) and for a side project in solid start (also vite). In those cases, the user navigates to another page between deploys and the client tries to load some JS and ultimately crashes because it can't find the files.

In the case of Qwik, if I'm not mistaken, this is far more than just a `Link` issue, it's for anything interactive.

I just initialized a new Qwik project to try this out. I did a build of the default project, then made one small change in an `onClick$` handler and did another build:

<img width="298" alt="image" src="https://github.com/BuilderIO/qwik/assets/174864/e1861783-7989-46ed-9055-4e6f94061efe">

This results in a massive diff in build output, which means that it's possible for many files to not load when they are required:

<img width="1179" alt="image" src="https://github.com/BuilderIO/qwik/assets/174864/f543d5ca-da9c-4c73-b9e7-9b4c51703695">

Is the solution to just keep static files around indefinitely on a CDN? For my full-time job we've resorted to showing a banner that says a new version of the app is available and that the user should refresh.

**Comment by knpwrs** [comment]
I just tried this out for real. After resetting to the default project state:

1. `npm run qwik add fastify`
2. `npm run build`
3. `npm run serve`
4. Visit the site
5. Make the change to the `onClick$` handler.
6. `npm run build`
7. `npm run serve`
8. Try clicking the time to celebrate button
9. Nothing happens
10. No celebration

**Comment by wmertens** [comment]
Yes, indeed all files should remain available for a while.

**Comment by iamriajul** [comment]
> It seems like a similar issues happens if the named export of a `routeAction$()` changes.

The filename is hash of code.

**Comment by iamriajul** [comment]
Currently, my deployment flow includes js files from the previous 100 builds. As I use CloudFront caching with stale-while-revalidate=1year, and it is far more likely for me to serve stale pages after many builds happened. You have deploy a lot of js files to backward compatible with stale pages cached by CDNs.

**Comment by gioboa** [comment]
Thanks @iamriajul that's a good solution 👏
I'm thinking about a custom Vite plugin that can read and include in the final bundle the previous build JS files during a new build.
This approach works with 2 consecutive builds but not with more than one. WDYT?

**Comment by iamriajul** [comment]
> This approach works with 2 consecutive builds but not with more than one. WDYT?

The `vite plugin solution` wouldn't solve the issue entirely, but will reduce it. And won't even work with adapters + Git Push deployment with providers like Vercel, Netlify etc.

**Comment by iamriajul** [comment]
I think there should be some kind of solution like `qwik-insights`. That can be a solution provided by Builder.io and open-source so anyone can host themselves like we can do with qwik-insights.

**Comment by iamriajul** [comment]
There will be a corresponding vite plugin that will connect to the Service to upload build files to the Builder.io server or Self-hosted server. 
So there will be always a backup of deployed files centrally managed in that Product (Sass maybe?)
When building a new build it fetches all the previous build files and places those inside the `dist/build` directory.

**Comment by gioboa** [comment]
I see, but that's something out of Qwik scope at the moment. So the long stale-while-revalidate approach is the best one. I think we can close this issue. WDYT? Thanks for your commitment in Qwik 👍

**Comment by iamriajul** [comment]
Before closing, the issue should be documented in the docs, so people can take their precautions as I have to think deeply and discover the issue and then take action.
Thanks.

**Comment by iamriajul** [comment]
I would really to know @mhevery 's thoughts on this issue and the solution / Sass I proposed. Thanks.

**Comment by gioboa** [comment]
In the Qwik Office Hours he told us that stale-while-revalidate is a good solution. Would you like to drop a PR for that?

**Comment by iamriajul** [comment]
> stale-while-revalidate

I don't see a direct relation to `stale-while-revalidate` with this issue. stale-while-revalidate is used for optimal caching when you have a CDN in front of your application, and I mentioned it because we use `stale-while-revalidate` for our application, and the likelihood of hitting a cached version of html page is way higher than a normal application setup.

If you're indicating things like using `stale-while-revalidate` with /build/ directory, then it is some kind of meaningful but that doesn't solve the issue because sometimes your some build files might not hit the CDN URL by user, for example when you have a button in a page where user doesn't interact frequently.

**Comment by elliotwaite** [comment]
@iamriajul, I have also been thinking about this issue. I like the idea of keeping the old files around to solve the file-loading issue, but there seems to be at least two other areas where version skew could cause issues: the client-server interface and the server-service interface.

**Client-server interface:**
This issue is described in the docs [here](https://qwik.builder.io/docs/server$/#how-does-server-work):
> Note: The server$() function must ensure that the server and client have the same version of the code executing. If there is a version skew the behavior is undefined and may result in an error. If version skew is a common problem then a more formal RPC mechanism should be used such as a tRPC or other library.

One potential solution to this would be to keep the entire container of the older server running and have a reverse proxy like Nginx decide which request should go to which server depending on something like a session cookie that stores which version of the code the client is running.

**Server-service interface:**
Even with the above solutions to the file-loading and the client-server issues, there is also the issue of the old server code no longer working correctly due to an external service having been changed when the new version was deployed. For example, a database migration could have been run that would make certain parts of the old server code no longer work as expected.

One potential solution to this database migration example would be to communicate with the older server before starting the database migration to let it know that a migration is going to occur. And then wherever the older server was going to communicate with the database, it could instead return an error message letting the user know that they need to refresh the page to update the app. Or for even more granularity, the older server could be told which tables are going to change so that it would only need to send an error message if it was going to interact with one of those tables.

So there seem to be at least three places where version skew could cause issues: file loading, the client-server interface, and the server-service interface. Keeping the old build files available for download seems to solve the file-loading issue, but I'd be interested to hear what others think about the other two areas where version skew could occur. I'd be open to brainstorming potential solutions.

And let me know if I should instead move this more general discussion of version skew to a different issue or over to Discord if you'd prefer to keep this discussion more focused on just the file-loading issue.

**Comment by iamriajul** [comment]
I've seen in the Laravel & Inertia community to force reload when there's such a version mismatch. 
Source: https://inertiajs.com/asset-versioning

**Comment by iamriajul** [comment]
I think it's better to move this problem into a new discussion as a whole.

**Comment by elliotwaite** [comment]
> I think it's better to move this problem into a new discussion as a whole.

Sounds good. I created a Discord post [here](https://discord.com/channels/842438759945601056/1192996848114348103).

**Comment by gioboa** [comment]
Closed by #5521
